src/net/protocol_agnostic/stream.rs

```diff
@@ -133,78 +136,101 @@ impl PaStream {
-                })
-                .or_else(|either| match either {
-                    Either::A((tcp_error, udp_connect)) => {
-                        udp_connect
-                            .map_err(move |utp_error| {
-                                PaRendezvousConnectError::AllProtocolsFailed {
-                                    tcp: Box::new(tcp_error),
-                                    utp: Box::new(utp_error),
-                                }
-                            })
-                            .into_boxed()
-                    }
-                    Either::B((utp_error, tcp_connect)) => {
-                        tcp_connect
-                            .map_err(move |tcp_error| {
-                                PaRendezvousConnectError::AllProtocolsFailed {
-                                    tcp: Box::new(tcp_error),
-                                    utp: Box::new(utp_error),
-                                }
-                            })
-                            .into_boxed()
-                    }
+        let handle = handle.clone();
+        let tcp_connect = {
+            TcpStream::rendezvous_connect(tcp_ch_1, &handle, p2p).map(PaStream::Tcp)
+        };
+        let udp_connect = {
+            UdpSocket::rendezvous_connect(utp_ch_1, &handle, p2p)
+                .map_err(UtpRendezvousConnectError::Rendezvous)
+                .and_then(move |(udp_socket, addr)| {
+                    trace!("udp rendezvous connect succeeded.");
+                    let (utp_socket, utp_listener) = {
+                        UtpSocket::from_socket(udp_socket, &handle).map_err(
+                            UtpRendezvousConnectError::IntoUtpSocket,
+                        )?
+                    };
+                    trace!("returning utp socket.");
+                    Ok((utp_socket, utp_listener, addr))
                 })
         };
 
+
         let ret = {
             pump_channels
-                .select2(connect)
-                .map_err(|either| match either {
-                    Either::A((err, _)) |
-                    Either::B((err, _)) => err,
-                })
-                .and_then(|either| match either {
-                    Either::A(((), connect)) => connect.into_boxed(),
-                    Either::B((stream, _)) => future::ok(stream).into_boxed(),
-                })
+            .while_driving(tcp_connect)
+            .while_driving(udp_connect)
+            .map_err(|((e, _tcp_connect), _udp_connect)| e)
+            .and_then(move |((their_pk, tcp_connect), udp_connect)| {
+                if our_pk > their_pk {
+                    let utp_connect = {
+                        udp_connect
+                        .and_then(|(utp_socket, _utp_listener, addr)| {
+                            utp_socket
+                                .connect(&addr)
+                                //.map_err(UtpRendezvousConnectError::UtpConnect)
```

Comment:
I'd say let's not keep commented out code - it's cryptic for people who did not write it.

Reply:
Another mistake that got fixed in a later commit. Sorry, it's just cumbersome to back-port small fixes when I've already layered other commits on top of the error. Unless you know of an easier way to do this?

Reply:
My workflow for that is

```
$ git checkout -b temp-branch
$ git checkout real-branch
$ git reset --hard commit-that-needs-to-be-changed
$ <make changes>
$ git commit --amend
$ git checkout temp-branch
$ git rebase --onto real-branch commit-that-needs-to-be-changed temp-branch
$ git checkout real-branch
$ git merge --ff-only temp-branch
$ git branch -d temp-branch
```

And then notice there was another spurious comment or something a few lines down and have to do it again.

Reply:
wow, indeed that looks like a lot of work.
Personally, I'm using git interactive rebase, it's described in here: https://git-scm.com/book/id/v2/Git-Tools-Rewriting-History
In a nutshell, it works smth like this:
```
$ git checkout real-branch  # just explicitly saying where I am :)
$ git rebase -i HEAD~5  # opens editor that allows me to change 5 commits back from now
$ < I choose commit i want to edit and git checksout this commit>
$ <make changes>
$ git commit --amend
$ git rebase --continue
```
I could demonstrate that over hangout, if you'd like ;)

Reply:
Thanks, I can copy what you've done there. I've never bothered to learn  `git rebase -i`, I guess I should because that's not the first nifty trick I've seen done with it.